presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java

```diff
@@ -263,29 +264,41 @@ private Page processColumnSource()
 
     private Page processPageSource()
     {
-        if (!finishing && !currentOutput.hasNext()) {
+        while (!finishing && !currentOutput.hasNext() && pageSource.isBlocked().isDone()) {
             Page page = pageSource.getNextPage();
 
             finishing = pageSource.isFinished();
             pageSourceMemoryContext.setBytes(pageSource.getSystemMemoryUsage());
 
-            if (page == null) {
-                currentOutput = EMPTY_PAGE_PROCESSOR_OUTPUT;
-            }
-            else {
+            if (page != null) {
                 // update operator stats
                 long endCompletedBytes = pageSource.getCompletedBytes();
                 long endReadTimeNanos = pageSource.getReadTimeNanos();
                 operatorContext.recordGeneratedInput(endCompletedBytes - completedBytes, page.getPositionCount(), endReadTimeNanos - readTimeNanos);
                 completedBytes = endCompletedBytes;
                 readTimeNanos = endReadTimeNanos;
 
-                currentOutput = pageProcessor.process(operatorContext.getSession().toConnectorSession(), page);
+                PageProcessorOutput output = pageProcessor.process(operatorContext.getSession().toConnectorSession(), page);
+                currentOutput.accept(output);
```

Comment:
You eagerly consume all `PageProcessorOutput` pages here. I don't think this is a good approach since `PageProcessorOutput` was written in such a way that it only produces new pages on `next()` call. By eagerly consuming pages you break this mechanism. This might be relevent when you append a lot of (heavy) columns to source page. In such case source page will be split into multiple pages, but you will compute all of them here and therefore reduce computations granuality.